templates/cluster-template.yaml

```
@@ -47,32 +47,37 @@ spec:
         net.bridge.bridge-nf-call-ip6tables = 1
         EOF
       - sysctl --system
-      - apt-get -y update
-      - DEBIAN_FRONTEND=noninteractive apt-get install -y apt-transport-https curl
-      - curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
-      - echo "deb https://apt.kubernetes.io/ kubernetes-xenial main" > /etc/apt/sources.list.d/kubernetes.list
-      - apt-get update -y
-      - TRIMMED_KUBERNETES_VERSION=$(echo {{ .kubernetesVersion }} | sed 's/\./\\./g' | sed 's/^v//')
-      - RESOLVED_KUBERNETES_VERSION=$(apt-cache policy kubelet | awk -v VERSION=$${TRIMMED_KUBERNETES_VERSION} '$1~ VERSION { print $1 }' | head -n1)
-      - apt-get install -y ca-certificates socat jq ebtables apt-transport-https cloud-utils prips containerd kubelet=$${RESOLVED_KUBERNETES_VERSION} kubeadm=$${RESOLVED_KUBERNETES_VERSION} kubectl=$${RESOLVED_KUBERNETES_VERSION}
+      - |
+        apt-get -y update
+        DEBIAN_FRONTEND=noninteractive apt-get install -y apt-transport-https curl
+        curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+        echo "deb https://apt.kubernetes.io/ kubernetes-xenial main" > /etc/apt/sources.list.d/kubernetes.list
+        apt-get update -y
+        TRIMMED_KUBERNETES_VERSION=$(echo {{ .kubernetesVersion }} | sed 's/\./\\./g' | sed 's/^v//')
+        RESOLVED_KUBERNETES_VERSION=$(apt-cache policy kubelet | awk -v VERSION=$${TRIMMED_KUBERNETES_VERSION} '$1~ VERSION { print $1 }' | head -n1)
+        apt-get install -y ca-certificates socat jq ebtables apt-transport-https cloud-utils prips containerd kubelet=$${RESOLVED_KUBERNETES_VERSION} kubeadm=$${RESOLVED_KUBERNETES_VERSION} kubectl=$${RESOLVED_KUBERNETES_VERSION}
       - systemctl daemon-reload
       - systemctl enable containerd
       - systemctl start containerd
-      - ping -c 3 -q {{ .controlPlaneEndpoint }} && echo OK || ip addr add {{ .controlPlaneEndpoint }} dev lo
+      - |
+        if [ -f "/run/kubeadm/kubeadm.yaml" ]; then
+          KVVERSION="v0.4.2"
+          ctr image pull ghcr.io/kube-vip/kube-vip:$${KVVERSION}
+          mkdir -p /etc/kubernetes/manifests/
+          ctr run --rm --net-host ghcr.io/kube-vip/kube-vip:$${KVVERSION} vip /kube-vip manifest pod --interface "lo" --vip "{{ .controlPlaneEndpoint }}" --controlplane --bgp --metal --metalKey "{{ .apiKey }}" --metalProjectID "b264ba74-f3bd-49f0-ae24-b0953765b7aa" | tee  /etc/kubernetes/manifests/kube-vip.yaml
```

This will expose the KUBE_VIP_VERSION as a variable to clusterctl to allow folks to easily override it if they wish.
```suggestion
          KVVERSION="${KUBE_VIP_VERSION:=v0.4.2}"
          ctr image pull ghcr.io/kube-vip/kube-vip:$${KVVERSION}
          mkdir -p /etc/kubernetes/manifests/
          ctr run --rm --net-host ghcr.io/kube-vip/kube-vip:$${KVVERSION} vip /kube-vip manifest pod --interface "lo" --vip "{{ .controlPlaneEndpoint }}" --controlplane --bgp --metal --metalKey "{{ .apiKey }}" --metalProjectID "b264ba74-f3bd-49f0-ae24-b0953765b7aa" | tee  /etc/kubernetes/manifests/kube-vip.yaml
```

---

Should this use `bond0` for the interface rather than `lo`?

---

It looks like projectID is being hardcoded here, it should use `${PROJECT_ID}` instead

---

The project ID is an accidental error in the template, I’ll fix that first thing in the morning along with making the kube-vip version configurable. 

As far as I’ve found in the kube-vip docs, the `lo` interface should be used:

[Note we bind the address to lo as we don't want multiple devices that have the same address on public interfaces.](https://kube-vip.chipzoller.dev/docs/installation/static/#bgp).

---

Ah, gotcha, thanks for the clarification.